aptos-move/framework/aptos-framework/sources/TokenCoinSwap.move

```diff
@@ -0,0 +1,231 @@
+// A basic utility module for swapping a fixed amount of coins with a fixed amount of tokens.
+module AptosFramework::TokenCoinSwap {
+    use Std::Signer;
+    use Std::Option::Option;
+    use AptosFramework::Table::{Self, Table};
+    use AptosFramework::Token::{Self, Token, TokenId};
+    use AptosFramework::Coin::{Self, Coin};
+    use AptosFramework::SimpleMap::{SimpleMap, Self};
+    use AptosFramework::Timestamp;
+
+    const ETOKEN_ALREADY_LISTED: u64 = 1;
+    const ETOKEN_LISTING_NOT_EXIST: u64 = 2;
+    const ETOKEN_NOT_IN_ESCROW: u64 = 3;
+    const ETOKEN_CANNOT_MOVE_OUT_OF_ESCROW_BEFORE_LOCKUP_TIME: u64 = 4;
+    const ETOKEN_MIN_PRICE_NOT_MATCH: u64 = 5;
+    const ETOKEN_AMOUNT_NOT_MATCH: u64 = 6;
+    const ENOT_ENOUGH_COIN: u64 = 7;
+
+    struct TokenCoinSwap<phantom CoinType> has store, drop {
+        token_amount: u64,
+        min_price_per_token: u64,
+    }
+
+    // stored at token owner's account
+    struct TokenListings<phantom CoinType> has key {
+        // key is the token id for swapping and value is the min price of target coin type.
+        listings: Table<TokenId, TokenCoinSwap<CoinType>>
```

Comment:
Tables are really only intended for large scale storage (many thousands+). Do we expect this to be large? Otherwise should be `SmallMap` perhaps?

Reply:
It is more likely to be very very few accounts have a really large number of listings and majority won't have many thousands

My current understanding is each entry in the table is stored separately. If we want to access the data separately, it is better to use Table. SimpleMap stores data in one chunk. Its use-case is for accessing multiple entries at the same time.

Reply:
Yes, that is exactly right (table is good for separate access per entry).  However, if this would be just a dozen entries or so, then a SimpleMap fetched together with account data would still be better. Yet as you don't know the scale ahead, Table is good.